Running Detox on Android

This issue compiles problems when running Detox on Android. 

**React Native version mismatch**

When running Detox on GitHub Actions, the app wouldn't start because of a `React Native version mismatch` error. This seems to be due to React Native bundler (cache?). 
Once the app is bundled, it works. A solution might be to pre-bundle the app on CI and do not launch bundler.
(See [Build and Install unsigned apk on device without the development server?](https://stackoverflow.com/questions/35283959/build-and-install-unsigned-apk-on-device-without-the-development-server))

👉 The workaround used is to run the release build. 

---

**Need to add a build type "releaseTest"**

```
> Task :app:transformClassesAndResourcesWithR8ForReleaseAndroidTest FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:transformClassesAndResourcesWithR8ForReleaseAndroidTest'.
> com.android.tools.r8.CompilationFailedException: Compilation failed to complete
```

[building androidTest apk gets R8 errors: “already has a mapping”](https://stackoverflow.com/questions/58717741/building-androidtest-apk-gets-r8-errors-already-has-a-mapping)

---

**Can't connect to bundler on real device**

When running locally the debug build on a real device, it won't connect to bundler until `adb reverse tcp:8081 tcp:8081`

---

**java.lang.ClassNotFoundException: com.facebook.react.devsupport.DevSupportManagerImpl**

When running Detox test (on real device + CI):
```
Requested enabled DevSupportManager, but DevSupportManagerImpl class was not found or could not be created
```

Fixed by disabling ProGuard minification

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -42,3 +42,47 @@ jobs:
         with:
           name: detox-screenshots-ios
           path: artifacts
+  ANDROID_EMULATOR_TESTS:
+    name: Android Emulator Tests
+    runs-on: macOS-latest
+    strategy:
+      matrix:
+        EMULATOR_API: [28]
+        EMULATOR_ARCH: [x86_64]
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-java@v1
+        with:
+          java-version: 1.8
+      - uses: actions/setup-node@v1
+        with:
+          node-version: '12.x'
+      - run: yarn install
+      - run: cp google-services.json.dist android/app/google-services.json
+      - name: Enable clear text traffic in release builds
+        run: |
+          brew install xmlstarlet
+          xmlstarlet ed --inplace --ps --insert "/manifest/application" --type attr -n "android:usesCleartextTraffic" -v "true" android/app/src/main/AndroidManifest.xml
+      - run: node node_modules/.bin/detox build -c android.emu.release
+        env:
+          googleMapsApiKey: ${{ secrets.googleMapsApiKey }}
+      - name: Execute emulator tests
+        uses: reactivecircus/android-emulator-runner@v2
+        with:
+          api-level: ${{ matrix.EMULATOR_API }}
+          target: google_apis
+          arch: ${{ matrix.EMULATOR_ARCH }}
+          profile: Nexus 6
+          # https://github.com/wix/Detox/blob/master/docs/Introduction.AndroidEmulatorsBestPractices.md#booting-an-emulator-via-command-line
+          emulator-options: -verbose -no-window -no-snapshot -noaudio -no-boot-anim -gpu swiftshader_indirect -camera-back emulated -camera-front emulated
+          disable-animations: true
+          script: |
+            bash -c "echo 'hw.lcd.height=2560' >> /Users/runner/.android/avd/test.avd/config.ini"
+            bash -c "echo 'hw.lcd.width=1440' >> /Users/runner/.android/avd/test.avd/config.ini"
+            node node_modules/.bin/detox test -c android.att.release --device-name=emulator-5554 --loglevel trace --take-screenshots all
+      - name: Upload screenshots artifacts
+        uses: actions/upload-artifact@v1
+        if: failure()
+        with:
+          name: detox-screenshots-android
+          path: artifacts

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,6 +1,7 @@
 env:
   global:
     - NODE_OPTIONS=--max-old-space-size=8192
+
 matrix:
   include:
   - language: node_js
@@ -10,71 +11,6 @@ matrix:
     script:
       - yarn lint
       - yarn test
-  - language: android
-    env: ANDROID_API_LEVEL=18
-    sudo: required
-    jdk: openjdk8
-    before_cache:
-      - rm -f  $HOME/.gradle/caches/modules-2/modules-2.lock
-      - rm -rf $HOME/.gradle/caches/*/plugin-resolution/
-    cache:
-      directories:
-        - "$HOME/.cache/yarn"
-        - "$HOME/.gradle/caches/"
-        - "$HOME/.gradle/wrapper/"
-    android:
-      components:
-        - tools
-        - platform-tools
-        - build-tools-28.0.3
-        - android-$ANDROID_API_LEVEL
-        - extra-google-google_play_services
-    before_install:
-      - yes | sdkmanager "platforms;android-28"
-      - nvm install 12
-    before_script:
-      - npm install -g yarn
-      - yarn config set cache-folder "$HOME/.cache/yarn"
-      - yarn install
-      - cp google-services.json.dist android/app/google-services.json
-    script:
-      - cd android && ./gradlew clean && ./gradlew assembleDebug --stacktrace
-  # https://medium.com/@fawwazyusran/how-to-travis-ci-working-for-android-api-level-28-2fde7292813c
-  - language: android
-    env: ANDROID_API_LEVEL=28
-    sudo: required
-    jdk: openjdk8
-    before_cache:
-      - rm -f  $HOME/.gradle/caches/modules-2/modules-2.lock
-      - rm -rf $HOME/.gradle/caches/*/plugin-resolution/
-    cache:
-      directories:
-        - "$HOME/.cache/yarn"
-        - "$HOME/.gradle/caches/"
-        - "$HOME/.gradle/wrapper/"
-    android:
-      components:
-        - tools
-        - platform-tools
-        - tools
-        - extra-android-m2repository
-        - extra-google-google_play_services
-      licenses:
-        - 'android-sdk-preview-license-52d11cd2'
-        - 'android-sdk-license-.+'
-        - 'google-gdk-license-.+'
-    before_install:
-      - touch $HOME/.android/repositories.cfg
-      - yes | sdkmanager "platforms;android-$ANDROID_API_LEVEL"
-      - yes | sdkmanager "build-tools;28.0.3"
-      - nvm install 12
-    before_script:
-      - npm install -g yarn
-      - yarn config set cache-folder "$HOME/.cache/yarn"
-      - yarn install
-      - cp google-services.json.dist android/app/google-services.json
-    script:
-      - cd android && ./gradlew clean && ./gradlew assembleDebug --stacktrace
 
 notifications:
   email: false

--- a/android/app/build.gradle
+++ b/android/app/build.gradle
@@ -180,9 +180,15 @@ android {
             signingConfig signingConfigs.debug
         }
         release {
-            minifyEnabled enableProguardInReleaseBuilds
-            proguardFiles getDefaultProguardFile("proguard-android.txt"), "proguard-rules.pro"
-            signingConfig signingConfigs.release
+            if (System.getProperty('minifyEnabled', 'yes') == 'yes') {
+                minifyEnabled enableProguardInReleaseBuilds
+            }
+            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro', "${rootProject.projectDir}/../node_modules/detox/android/detox/proguard-rules-app.pro"
+            if (System.getProperty('useDebugCertificate', 'no') == 'yes') {
+                signingConfig signingConfigs.debug
+            } else {
+                signingConfig signingConfigs.release
+            }
         }
     }
     // applicationVariants are e.g. debug, release

--- a/e2e/01_dispatch.spec.js
+++ b/e2e/01_dispatch.spec.js
@@ -7,9 +7,12 @@ describe('Dispatch', () => {
     await device.reloadReactNative()
 
     // https://github.com/wix/Detox/issues/1371
-    // await device.setLocation(48.856613, 2.352222);
-    const { stdout, stderr } = await exec("idb list-targets | grep -n 'Booted'");
-    await exec(`idb set-location --udid ${device._deviceId} 48.856613 2.352222`);
+    if (device.getPlatform() === 'ios') {
+      const { stdout, stderr } = await exec("idb list-targets | grep -n 'Booted'");
+      await exec(`idb set-location --udid ${device._deviceId} 48.856613 2.352222`);
+    } else {
+      await device.setLocation(48.856613, 2.352222);
+    }
   })
 
   it('should be able to create task', async () => {
@@ -45,7 +48,10 @@ describe('Dispatch', () => {
 
     await element(by.id('loginUsername')).typeText('bot_1')
     await element(by.id('loginPassword')).typeText('bot_1')
-    await element(by.id('loginSubmit')).tap()
+
+    try {
+      await element(by.id('loginSubmit')).tap()
+    } catch (e) {}
 
     await element(by.id('messengerTabList')).tap()
 

--- a/e2e/03_registration.spec.js
+++ b/e2e/03_registration.spec.js
@@ -21,16 +21,19 @@ describe('Registration', () => {
 
     const timestamp = moment().format('X')
 
-    await element(by.id('registerForm.email')).typeText(`dev+${timestamp}@coopcycle.org`)
-    await element(by.id('registerForm.username')).typeText(`dev${timestamp}`)
+    await element(by.id('registerForm.email')).typeText(`dev+${timestamp}@coopcycle.org\n`)
+    await element(by.id('registerForm.username')).typeText(`dev${timestamp}\n`)
     // https://github.com/wix/detox/issues/916
-    await element(by.id('registerForm.password')).replaceText('12345678')
-    await element(by.id('registerForm.passwordConfirmation')).replaceText('12345678')
+    await element(by.id('registerForm.password')).replaceText('12345678\n')
+    await element(by.id('registerForm.passwordConfirmation')).replaceText('12345678\n')
 
-    await element(by.id('registerForm.givenName')).typeText('John')
-    await element(by.id('registerForm.familyName')).typeText('Doe')
+    await element(by.id('registerForm.givenName')).typeText('John\n')
+    await element(by.id('registerForm.familyName')).typeText('Doe\n')
 
-    await element(by.id('submitRegister')).tap()
+    // As we are using "\n", the form may have been submitted yet
+    try {
+        await element(by.id('submitRegister')).tap()
+    } catch (e) {}
 
     await expect(element(by.id('registerCheckEmail'))).toBeVisible()
   })

--- a/e2e/04_checkout.spec.js
+++ b/e2e/04_checkout.spec.js
@@ -41,7 +41,7 @@ describe('Checkout', () => {
     await element(by.id('menuItem:0:1')).tap()
     await element(by.id('menuItem:0:2')).tap()
 
-    await waitFor(element(by.id('cartSubmit'))).toExist().withTimeout(5000)
+    await waitFor(element(by.id('cartSubmit'))).toBeVisible().withTimeout(5000)
     await element(by.id('cartSubmit')).tap()
 
     await expect(element(by.id('cartSummarySubmit'))).toBeVisible()
@@ -50,9 +50,12 @@ describe('Checkout', () => {
     await expect(element(by.id('loginUsername'))).toBeVisible()
     await expect(element(by.id('loginPassword'))).toBeVisible()
 
-    await element(by.id('loginUsername')).typeText('user_1')
-    await element(by.id('loginPassword')).typeText('user_1')
-    await element(by.id('loginSubmit')).tap()
+    await element(by.id('loginUsername')).typeText('user_1\n')
+    await element(by.id('loginPassword')).typeText('user_1\n')
+
+    try {
+      await element(by.id('loginSubmit')).tap()
+    } catch (e) {}
 
     await expect(element(by.id('checkoutTelephone'))).toBeVisible()
     await expect(element(by.id('moreInfosSubmit'))).toBeVisible()
@@ -68,7 +71,8 @@ describe('Checkout', () => {
 
     await element(by.id('creditCardNumber')).typeText('4242424242424242')
     await element(by.id('creditCardExpiry')).typeText('1221')
-    await element(by.id('creditCardCvc')).typeText('123')
+    // Add "\n" to make sure keyboard is hidden
+    await element(by.id('creditCardCvc')).typeText('123\n')
 
     await element(by.id('creditCardSubmit')).tap()
 

--- a/e2e/init.js
+++ b/e2e/init.js
@@ -4,7 +4,7 @@ const adapter = require('detox/runners/jest/adapter');
 const specReporter = require('detox/runners/jest/specReporter');
 
 // Set the default timeout
-jest.setTimeout(120000);
+jest.setTimeout(240000);
 jasmine.getEnv().addReporter(adapter);
 
 // This takes care of generating status logs on a per-spec basis. By default, jest only reports at file-level.

--- a/e2e/utils.js
+++ b/e2e/utils.js
@@ -6,8 +6,13 @@ export const connectToDemo = async () => {
   await expect(element(by.id('moreServerOptions'))).toBeVisible()
   await element(by.id('moreServerOptions')).tap()
 
-  await element(by.id('customServerURL')).typeText('demo.coopcycle.org')
-  await element(by.id('submitCustomServer')).tap()
+  await element(by.id('customServerURL')).typeText('demo.coopcycle.org\n')
+
+  try {
+    // We deliberately add "\n" to hide the keyboard
+    // The tap below shouldn't be necessary
+    await element(by.id('submitCustomServer')).tap()
+  } catch (e) {}
 
 }
 
@@ -17,9 +22,13 @@ export const authenticateWithCredentials = async (username, password) => {
   await element(by.id('menuBtn')).tap()
   await element(by.id('drawerAccountBtn')).tap()
 
-  await element(by.id('loginUsername')).typeText(username)
-  await element(by.id('loginPassword')).typeText(password)
-  await element(by.id('loginSubmit')).tap()
+  await element(by.id('loginUsername')).typeText(`${username}\n`)
+  await element(by.id('loginPassword')).typeText(`${password}\n`)
+
+  // As we are using "\n", the form may have been submitted yet
+  try {
+      await element(by.id('loginSubmit')).tap()
+  } catch (e) {}
 
 }
 

--- a/package.json
+++ b/package.json
@@ -146,11 +146,16 @@
       },
       "android.emu.release": {
         "binaryPath": "android/app/build/outputs/apk/release/app-release.apk",
-        "build": "cd android && ./gradlew assembleRelease assembleAndroidTest -DtestBuildType=release --warning-mode all --stacktrace && cd ..",
+        "build": "cd android && ./gradlew assembleRelease assembleAndroidTest -DtestBuildType=release -DuseDebugCertificate=yes -DminifyEnabled=no --warning-mode all --stacktrace && cd ..",
         "type": "android.emulator",
         "device": {
           "avdName": "test"
         }
+      },
+      "android.att.release": {
+        "binaryPath": "android/app/build/outputs/apk/release/app-release.apk",
+        "build": "cd android && ./gradlew assembleRelease assembleAndroidTest -DtestBuildType=release -DuseDebugCertificate=yes -DminifyEnabled=no --warning-mode all --stacktrace && cd ..",
+        "type": "android.attached"
       }
     },
     "test-runner": "jest"

